bitcoin/src/bip32.rs

```diff
@@ -718,19 +788,22 @@ impl Xpub {
     /// the internal public key representation.
     pub fn to_x_only_pub(self) -> XOnlyPublicKey { XOnlyPublicKey::from(self.public_key) }
 
-    /// Attempts to derive an extended public key from a path.
+    /// Derives an extended public key from a path.
     ///
     /// The `path` argument can be any type implementing `AsRef<ChildNumber>`, such as `DerivationPath`, for instance.
     pub fn derive_pub<C: secp256k1::Verification, P: AsRef<[ChildNumber]>>(
         &self,
         secp: &Secp256k1<C>,
         path: &P,
-    ) -> Result<Xpub, Error> {
+    ) -> Xpub {
         let mut pk: Xpub = *self;
         for cnum in path.as_ref() {
-            pk = pk.ckd_pub(secp, *cnum)?
+            // convert the cnum into NormalChildNumber
+            let cnum: NormalChildNumber =
+                NormalChildNumber::try_from(*cnum).expect("cnum is not hardened");
```

Comment:
This is not what "prevent by type system" means. There should be no panic and `P` needs to be `AsRef<[NormalChildNumber]>` instead.

Though now I'm wondering why we're not using `IntoIterator<Item=(Normal)ChildNumber>` instead.

Reply:
Yes, I agree. I am trying the `AsRef<[NormalChildNumber]>` route. But I didn't get the implicit conversions correct yet.
I'll try harder 😄.